usersession: add systemd user instance service control to user session agent

jhenstridge opened the pull request:
This PR adds an API to the user session agent to control user instance systemd services.  It only aims to provide the features needed to manage user daemons, so covers `daemon-reload`, `start`, and `stop` (enabling and disabling services is done through the "global" `/etc/systemd/user` directory, so doesn't need access to the user instance).

There is a single `/v1/services` end point that accepts POST requests with a JSON body.  The `action` member in the body specifies which action should be performed, and the `services` member provides the list of services for `start`/`stop`.  As a restriction, the service names must start with `snap.`, to prevent control of non-snap services.

At present, this doesn't have an option to force kill services for e.g. daemons using weird kill modes.  That could probably be added later as an additional argument in the post body.

zyga commented on usersession/agent/rest_api.go:
```diff
@@ -46,3 +59,129 @@ func sessionInfo(c *Command, r *http.Request) Response {
+	if err != nil && systemd.IsTimeout(err) {
+		// ignore errors for kill; nothing we'd do differently at this point
+		sysd.Kill(service, "TERM", "")
+		time.Sleep(killWait)
```
Should we wait unconditionally? If the wait time is a bit larger then this will just wait even though the service may have reacted faster. I also wonder if we really need this. Systemd models behaviour for stopping services, along with any timeouts and cleanup necessary.

What we might do instead is kill stopping fails.

jhenstridge replied:
I adapted this code from `stopService` in `wrappers/services.go`.  Looking at the documentation in systemd.kill(5), it isn't obvious to me that this is needed either.

jhenstridge replied:
We're already forgoing support for the oldest systems here: Ubuntu 14.04 and CentOS/RHEL 7 don't instantiate a user mode systemd, so this code would never run on those releases.  I wonder if that raises the bar high enough that we can just rely on `stop` being reliable?

zyga commented on usersession/agent/rest_api.go:
```diff
@@ -46,3 +59,129 @@ func sessionInfo(c *Command, r *http.Request) Response {
+	// Refuse to start non-snap services
+	for _, service := range inst.Services {
+		if !strings.HasPrefix(service, "snap.") {
+			return InternalError("cannot start service %v", service)
```
```suggestion
			return InternalError("cannot start non-snap service %v", service)
```

jhenstridge replied:
Good idea.  I've made the same change to the error in `serviceStop`.

zyga commented on usersession/agent/rest_api.go:
```diff
@@ -38,6 +46,11 @@ var (
+
+	servicesCmd = &Command{
+		Path: "/v1/services",
+		POST: postServices,
```
Do we need any kind of authentication here? Do we need to enforce that the user making the request is the same as the user running the user session daemon?

jhenstridge replied:
The intended audience for this API is root.  Other users should not be able to access the the session agent due to the permissions on `$XDG_RUNTIME_DIR`.

On top of this, PR #7197 adds a second layer of protection by performing an `SO_PEERCRED` check on new connections.  If we're only performing actions on behalf of the same user or root, we don't need the complication of snapd's per-user authentication checks.

jdstrand replied:
FWIW, PR 7555 addresses my concerns about 'another line of defense' since with strict mode, apparmor will ensure snaps can't interfere with the socket. For non-snaps, the socket is protected via DAC so other non-root users can't access it and for the session user, the API provides no additional privilege. What is left is partial confinement systems where a snap might have access to the socket [due to a lack of unix mediation patches](https://github.com/snapcore/snapd/pull/7555#discussion_r354452340). From my pov, that is not a blocker, but there is an opportunity for a future hardening enhancement.

zyga commented on usersession/agent/rest_api.go:
```diff
@@ -46,3 +59,129 @@ func sessionInfo(c *Command, r *http.Request) Response {
+	if err := decoder.Decode(&inst); err != nil {
+		return BadRequest("cannot decode request body into service instruction: %v", err)
+	}
+	impl := serviceInstructionDispTable[inst.Action]
```
Do we need to grab a lock for this? If several instructions are posted concurrently I think we should lock around either globally (per user) or per snap (and refuse to issue commands that would affect multiple snaps in one go).

jhenstridge replied:
That seems reasonable.  I've added a global lock to serialise the final final portion of `postServices`.

jdstrand replied:
@zyga - I believe @jhenstridge change should address your concerns?

pedronis commented:
@chipaca this should be reviewed and integrated before #5822

chipaca commented:
@pedronis ack

jhenstridge commented on tests/main/snap-session-agent-service-control/task.yaml:
```diff
@@ -0,0 +1,88 @@
+        -d '{"action": "stop", "services": ["snap.test-service.service"]}' \
+        http://localhost/v1/services | MATCH "HTTP/1.1 200 OK"
+    systemctl_user show --property=ActiveState snap.test-service.service |
+        MATCH ActiveState=inactive
```
I've created PR #7555 with a spread test demonstrating this for strict confinement systems.  I agree that we might want to add further security checks for partial confinement systems, but it probably wouldn't have a meaningful impact on security even with this new session agent API.

If the app can connect to the session bus (either because there is no file level AppArmor protection, or because it has connected one of the interfaces that allows it), then it will be able to control the user instance of systemd directly.  At that point, it will be able to do everything this API allows and more.

jdstrand replied:
Thank you for PR 7555. I left my review here: https://github.com/snapcore/snapd/pull/7555#pullrequestreview-334233943 (approved)

chipaca commented:
give the pr has conflicts, maybe fix the nits here as you'll have to push it through spread again... :-)

pedronis commented:
@jhenstridge if you could look at conflicts and previous feedback here, I then could do my own pass on it

jhenstridge commented:
I've fixed the conflicts, and also removed the code to explicitly send TERM and KILL signals (as @zyga suggested earlier in the review).  The failing CI checks appear to be unrelated to the contents of this branch.

According to the systemd.kill(5) man page, systemd defaults to sending SIGKILL to any processes remaining in the cgroup after the timeout unless the service file sets SendSIGKILL=no.  Snapd never sets this in its service files, so the manual kill calls should never be necessary.  It sounds like they were added for the benefit of ancient buggy systemds, but this code will only run on systems with a functional user mode systemd instance.  That alone weeds out the really old systems.

pedronis commented on usersession/agent/rest_api.go:
```diff
@@ -46,3 +60,123 @@ func sessionInfo(c *Command, r *http.Request) Response {
+	if len(startErrors) != 0 {
+		for _, service := range started {
+			if err := sysd.Stop(service, stopTimeout); err != nil {
+				startErrors[service] = err.Error()
```
just recording these errors as start errors is a bit odd,  maybe we need 2 maps so that the client can decide how to treat them,  wrappers.StartServices treats them differently than base errors for examples

jhenstridge replied:
I've separate out the errors into `start-errors` and `stop-errors` dictionaries, and updated the "stop" action to use use similar nesting in the error value for consistency.  This should also make it possible to add more information to the error later if we find a need.

pedronis commented on usersession/agent/rest_api.go:
```diff
@@ -38,6 +47,11 @@ var (
 	}
+
+	servicesCmd = &Command{
+		Path: "/v1/services",
```
/services makes one think that is high-level and a deals in snap services, instead of systemd units.  Maybe we should be consistent with other naming here and call this /service-control

jhenstridge replied:
Good idea.  I've moed it to `/v1/service-control` now.

pedronis commented:
@jhenstridge seems there are still some previous unadressed feedback also this will need a 2nd pass by @jdstrand

jhenstridge commented on usersession/agent/rest_api.go:
```diff
@@ -46,3 +60,134 @@ func sessionInfo(c *Command, r *http.Request) Response {
+	killWait    = 5 * time.Second
+)
+
+func stopOneService(service string, sysd systemd.Systemd) error {
```
I ended up removing this helper function, on the basis that systemd versions providing a user instance are new enough not to need the work-arounds to kill services.  The call sites for this helper now use the generic `sysd.Stop` method directly, and have the appropriate guard.

pedronis commented:
@jhenstridge is this ready for a re-review ?

jhenstridge commented on tests/main/snap-session-agent-service-control/task.yaml:
```diff
@@ -0,0 +1,88 @@
+    if snap list test-snapd-curl; then
+        snap remove test-snapd-curl
+    fi
+    systemctl stop "user@${TEST_UID}.service"
```
If the `systemctl stop` command fails, doesn't that mean we are in an unknown state though?  It would mean we've left behind a running user session for the `test` user that could interfere with future tests.  If we somehow failed before starting `user@1234.service`, then the `stop` command is a no-op.

The only places I can see the `systemctl stop ... || true` pattern used in a restore section is for the `proxy` and `proxy-no-core` tests, which dynamically create the unit within the test body.  Here, the unit will always exist.

jhenstridge commented:
I think this is ready for re-review, yes.

pedronis commented:
@jhenstridge  did you see my question?

>  if I look at parallel code in wrappers etc, there is code to deal with sockets and timers as well as part of wrappers.StartServices itself etc, what's the plan about those for user session services? will the caller explode those and ask for them explicitly?

jhenstridge commented:
Sorry, I forgot to respond to that.  I don't think there is any need for special socket/timer handling at the session agent level.

These routines can't be user daemon equivalents to the high level `StartServices`/`StopServices` routines found in the wrappers package.  When we enable or disable user timers/sockets/services, that should happen at the global level (otherwise the services won't be enabled for users who are not currently logged in).

It's only the start/stop/daemon-reload calls in those high level routines that need to be proxied to the user session, and those calls don't care whether the unit is a service or a socket.

jdstrand commented on usersession/agent/rest_api_test.go:
```diff
@@ -74,3 +95,265 @@ func (s *restSuite) TestSessionInfo(c *C) {
+	c.Assert(agent.ServiceControlCmd.GET, IsNil)
+	c.Check(agent.ServiceControlCmd.PUT, IsNil)
+	c.Check(agent.ServiceControlCmd.POST, NotNil)
+	c.Check(agent.ServiceControlCmd.DELETE, IsNil)
```
Maybe also add HEAD and POSTFORM?

jhenstridge replied:
There are no HEAD or POSTFORM members on the command struct.

zyga commented:
Is this waiting on anyone's review? Can we just merge it?
